### PR TITLE
Support for zoom security improvements.

### PR DESCRIPTION
**Pull request by Tekkon**
This PR relates to this issue: https://github.com/hintmedia/zoom_rb/issues/401

### CODE REVIEW

**Comment by joelmichael** [comment]
Thanks for this. I just want you to know that I've seen it and will review it more properly soon

File: lib/zoom/actions.rb [hunk 1]
```diff
@@ -26,31 +26,40 @@ def self.determine_request_options(client, oauth)
...
+        required, permitted, oauth, args_to_params, headers =
+          options.values_at :require, :permit, :oauth, :args_to_params, :headers
         required = Array(required) unless required.is_a?(Hash)
         permitted = Array(permitted) unless permitted.is_a?(Hash)
 
         define_method(name) do |*args|
           path_keys = Zoom::Actions.extract_path_keys(path)
-          params = Zoom::Params.new(Utils.extract_options!(args))
+          params = Utils.extract_options!(args)
+          args_to_params&.each { |key, value| params[value] = params.delete key if params[key] }
```
Comment (joelmichael) [inline lib/zoom/actions.rb:52]: Is the purpose of `args_to_params` to essentially allow backwards compatibility between the old API (using `auth_code`) and the correct Zoom API (using `code`)? Also `auth_token` to `token`. I might be more inclined to just update the `OAuth` class to change its calls to use the new (correct) format (`code` and `token`) and also require users to change their usage of the gem to meet the new format, with a Breaking Change noted in the Changelog. But I admit that could cause problems for many users if they don't see it. Perhaps the `OAuth` class could allow either argument as input but only use the new ones in its API calls to Zoom.

File: lib/zoom/actions.rb [hunk 2]
```diff
@@ -26,31 +26,40 @@ def self.determine_request_options(client, oauth)
...
+    def self.make_request(args)
+      client, method, parsed_path, params, request_options, oauth =
+        args.values_at :client, :method, :parsed_path, :params, :request_options, :oauth
       case method
       when :get
         request_options[:query] = params
       when :post, :put, :patch
-        request_options[:body] = params.to_json
+        request_options[:body] =
+          oauth ? URI.encode_www_form(params.to_a) : params.to_json
```
Comment (joelmichael) [inline lib/zoom/actions.rb:37]: This seems right 👍

File: lib/zoom/actions.rb [hunk 3]
```diff
@@ -26,31 +26,40 @@ def self.determine_request_options(client, oauth)
       end
     end
 
-    def self.make_request(client, method, parsed_path, params, request_options)
+    def self.make_request(args)
+      client, method, parsed_path, params, request_options, oauth =
+        args.values_at :client, :method, :parsed_path, :params, :request_options, :oauth
```
Comment (joelmichael) [inline lib/zoom/actions.rb:31]: I'm not sure how this is beneficial over the old list of arguments in the function definition. It seems like a refactor that may be outside the scope of this change.

File: lib/zoom/client.rb [hunk 4]
```diff
@@ -35,14 +35,20 @@ def headers
...
-    def request_headers
+    def bearer_authorization_header
       {
         'Authorization' => "Bearer #{access_token}"
-      }.merge(headers)
+      }
+    end
+
+    def request_headers
+      bearer_authorization_header.merge(headers)
```
Comment (joelmichael) [inline lib/zoom/client.rb:51]: This seems like a refactor that is outside the scope of this change. I would think we should just use `bearer_authorization_header` and not use `request_headers` at all. I agree it makes sense to differentiate between the oauth headers and the "standard" headers more precisely.

**Review by joelmichael — COMMENTED** [review]
I think this is very helpful work that will enable users of this gem to address the May 15, 2022 deadline for using the new OAuth format. The sooner we get this out and allow users to upgrade the better. However I have a few comments and would appreciate if you have any thoughts on them. I might be willing to go with this as-is however just because it's so important.

File: lib/zoom/actions.rb (hunk 1 above)
Comment (benjaminwood) [reply lib/zoom/actions.rb:52]: Perhaps we keep backwards compatibility for now but add a deprecation warning and plan to remove it in the next major version change?

Just suggesting a 3rd option - your call though @joelmichael.

File: lib/zoom/actions.rb (hunk 3 above)
Comment (Tekkon) [reply lib/zoom/actions.rb:31]: @joelmichael I just considered, that more than five arguments is too much, that's why I enclosed it into just one hash argument.

File: lib/zoom/client.rb (hunk 4 above)
Comment (Tekkon) [reply lib/zoom/client.rb:51]: @joelmichael Do you mean use `bearer_authorization_header` without merging it with `headers`?

**Comment by Tekkon** [comment]
I also added support for code_verifier parameter for access_tokens call. Please, have a look.

**Comment by joelmichael** [comment]
I am going to merge this as-is today and cut a new minor version of the gem. I appreciate your work on this and your patience in it being merged.
